File: editors/vscode/src/frontmatter.ts
import * as vscode from "vscode";
import { EvtSpec } from "./spec";

export const EVT_SELECTOR: vscode.DocumentSelector = { language: "evt" };

// ---- frontmatter parsing --------------------------------------------------

interface Frontmatter {
  startLine: number; // the opening ---
  endLine: number;   // the closing --- (exclusive of body)
}

// The leading `---` … `---` block, or null if the file has no signature.
function frontmatterRange(doc: vscode.TextDocument): Frontmatter | null {
  if (doc.lineCount === 0 || doc.lineAt(0).text.trim() !== "---") return null;
  for (let i = 1; i < doc.lineCount; i++) {
    if (doc.lineAt(i).text.trim() === "---") return { startLine: 0, endLine: i };
  }
  return null; // unterminated: treated as all-body by the runtime
}

function inFrontmatter(doc: vscode.TextDocument, line: number): Frontmatter | null {
  const fm = frontmatterRange(doc);
  return fm && line > fm.startLine && line < fm.endLine ? fm : null;
}

// The nearest top-level `key:` above `line` (the section the cursor is editing under).
function sectionAt(doc: vscode.TextDocument, fm: Frontmatter, line: number): string | null {
  for (let i = line; i > fm.startLine; i--) {
    const m = /^([A-Za-z_]+):\s*(.*)$/.exec(doc.lineAt(i).text);
    if (m) return m[1];
  }
  return null;
}

function isNodeScript(doc: vscode.TextDocument): boolean {
  return doc.fileName.endsWith(".node.evt");
}

// The capability apis this file declared in `apis:` (what the sandbox actually exposes).
export function declaredApis(doc: vscode.TextDocument): Set<string> {
  const fm = frontmatterRange(doc);
  return fm ? listedUnder(doc, fm, "apis") : new Set();
}

function hooksFor(doc: vscode.TextDocument, spec: EvtSpec): string[] {
  return isNodeScript(doc) ? spec.nodeHooks : spec.systemHooks;
}

// Values already listed under a section (so completion can skip them).
function listedUnder(doc: vscode.TextDocument, fm: Frontmatter, section: string): Set<string> {
  const out = new Set<string>();
  let inSection = false;
  for (let i = fm.startLine + 1; i < fm.endLine; i++) {
    const text = doc.lineAt(i).text;
    const key = /^([A-Za-z_]+):/.exec(text);
    if (key) { inSection = key[1] === section; continue; }
    if (!inSection) continue;
    const item = /^\s*-\s*([^:#\s]+)/.exec(text);
    if (item) out.add(item[1]);
  }
  return out;
}

// ---- completion -----------------------------------------------------------

function completionProvider(spec: () => EvtSpec): vscode.CompletionItemProvider {
  return {
    provideCompletionItems(doc, pos) {
      const fm = inFrontmatter(doc, pos.line);
      if (!fm) return undefined;
      const s = spec();
      const line = doc.lineAt(pos.line).text;
      const item = /^(\s*)-\s*(\S*)$/.exec(line.slice(0, pos.character));

      // A list item under a section -> complete that section's values.
      if (item) {
        const section = sectionAt(doc, fm, pos.line);
        if (section === "apis") {
          const used = listedUnder(doc, fm, "apis");
          return s.apis.filter((a) => !used.has(a)).map((a) => {
            const c = new vscode.CompletionItem(a, vscode.CompletionItemKind.Module);
            c.detail = "EvaLuate capability api";
            c.documentation = new vscode.MarkdownString(
              s.apiNotes[a] ?? `Members: ${(s.apiMembers[a] ?? []).map((m) => `\`${a}.${m}\``).join(", ") || "(node)"}`,
            );
            return c;
          });
        }
        if (section === "register") {
          const used = listedUnder(doc, fm, "register");
          return hooksFor(doc, s).filter((h) => !used.has(h)).map((h) => {
            const c = new vscode.CompletionItem(h, vscode.CompletionItemKind.Event);
            c.detail = isNodeScript(doc) ? "node lifecycle hook" : "system lifecycle hook";
            return c;
          });
        }
        return undefined; // config/scenes/assets/returns: free text
      }

      // Start of a line at the top level -> complete frontmatter keys.
      if (/^\s*[A-Za-z_]*$/.test(line.slice(0, pos.character)) && !/^\s+/.test(line)) {
        const present = new Set<string>();
        for (let i = fm.startLine + 1; i < fm.endLine; i++) {
          const k = /^([A-Za-z_]+):/.exec(doc.lineAt(i).text);
          if (k) present.add(k[1]);
        }
        return s.frontmatterKeys.filter((k) => !present.has(k)).map((k) => {
          const c = new vscode.CompletionItem(k, vscode.CompletionItemKind.Property);
          c.insertText = new vscode.SnippetString(`${k}:\n - $0`);
          c.detail = "frontmatter key";
          return c;
        });
      }
      return undefined;
    },
  };
}

// ---- diagnostics ----------------------------------------------------------

function validate(doc: vscode.TextDocument, s: EvtSpec, sink: vscode.DiagnosticCollection): void {
  if (doc.languageId !== "evt") return;
  const fm = frontmatterRange(doc);
  if (!fm) { sink.delete(doc.uri); return; }

  const diags: vscode.Diagnostic[] = [];
  const validHooks = new Set(hooksFor(doc, s));
  const validApis = new Set(s.apis);
  const validKeys = new Set(s.frontmatterKeys);
  let section: string | null = null;

  for (let i = fm.startLine + 1; i < fm.endLine; i++) {
    const text = doc.lineAt(i).text;
    const key = /^([A-Za-z_]+):/.exec(text);
    if (key) {
      section = key[1];
      if (!validKeys.has(section)) {
        diags.push(diag(i, key[1], 0, `Unknown frontmatter key '${section}'. Valid: ${[...validKeys].join(", ")}.`,
          vscode.DiagnosticSeverity.Warning));
      }
      continue;
    }
    const item = /^(\s*-\s*)([^:#\s]+)/.exec(text);
    if (!item) continue;
    const value = item[2];
    const col = item[1].length;
    if (section === "apis" && !validApis.has(value) && !value.startsWith("godot:")) {
      diags.push(diag(i, value, col, `Unknown api '${value}'. Valid: ${[...validApis].join(", ")}. (It must be declared to be usable.)`,
        vscode.DiagnosticSeverity.Warning));
    } else if (section === "register" && !validHooks.has(value)) {
      diags.push(diag(i, value, col,
        `Unknown ${isNodeScript(doc) ? "node" : "system"} hook '${value}'. Valid: ${[...validHooks].join(", ")}.`,
        vscode.DiagnosticSeverity.Warning));
    }
  }

  // Sandbox fidelity: a capability api used in the body but NOT declared in `apis:` is
  // absent at runtime (the sandbox only exposes declared apis), so the script would error.
  // Flag it here the same way the body's completion hides undeclared apis.
  for (const d of undeclaredApiUses(doc, fm.endLine, validApis, declaredApis(doc))) diags.push(d);

  sink.set(doc.uri, diags);
}

// Find body references to a capability api that the frontmatter did not declare. A use is
// `name.` / `name:` / `name(`; skipped if the body locally rebinds the name (`local name`
// or `name =`), which would shadow the (absent) global.
function undeclaredApiUses(
  doc: vscode.TextDocument, bodyStart: number, allApis: Set<string>, declared: Set<string>,
): vscode.Diagnostic[] {
  const out: vscode.Diagnostic[] = [];
  // Strip line comments once; an api named in a comment isn't a real use.
  const code: string[] = [];
  for (let i = 0; i < doc.lineCount; i++) code[i] = i > bodyStart ? doc.lineAt(i).text.replace(/--.*$/, "") : "";
  const body = code.join("\n");

  // An api the body locally rebinds (`local save` / `save = ...`) shadows the absent global,
  // so it isn't a sandbox violation — exclude it across the WHOLE body, not line-by-line.
  const undeclared = [...allApis].filter(
    (a) => !declared.has(a) && !new RegExp(`\\blocal\\s+${a}\\b|(^|[^.\\w])${a}\\s*=[^=]`).test(body),
  );
  if (undeclared.length === 0) return out;

  for (let i = bodyStart + 1; i < doc.lineCount; i++) {
    for (const api of undeclared) {
      const use = new RegExp(`(^|[^.:\\w])(${api})\\s*[.:(]`, "g");
      let m: RegExpExecArray | null;
      while ((m = use.exec(code[i])) !== null) {
        const col = m.index + m[1].length;
        out.push(diag(i, api, col, `'${api}' is used but not declared in 'apis:'. ` +
          `The sandbox only exposes declared apis, so this is nil at runtime. Add '${api}' to 'apis:'.`,
          vscode.DiagnosticSeverity.Warning));
      }
    }
  }
  return out;
}

function diag(line: number, token: string, col: number, message: string, severity: vscode.DiagnosticSeverity): vscode.Diagnostic {
  const range = new vscode.Range(line, col, line, col + token.length);
  const d = new vscode.Diagnostic(range, message, severity);
  d.source = "evt";
  return d;
}

// ---- hover ----------------------------------------------------------------

function hoverProvider(spec: () => EvtSpec): vscode.HoverProvider {
  const keyDocs: Record<string, string> = {
    config: "TOML files exposed as `config.<section>.*` (read live; hot-reloads).",
    apis: "Capability apis added to the sandbox. Only what you list here is reachable.",
    register: "Lifecycle hooks this script handles — each needs a matching global `function`.",
    returns: "The narrowed module contract for `require` consumers.",
    assets: "Files watched for hot-reload alongside the script.",
    scenes: "System scripts only: restrict hooks to these scenes (omit ⇒ global).",
  };
  return {
    provideHover(doc, pos) {
      const fm = inFrontmatter(doc, pos.line);
      if (!fm) return undefined;
      const range = doc.getWordRangeAtPosition(pos, /[A-Za-z_][A-Za-z0-9_]*/);
      if (!range) return undefined;
      const word = doc.getText(range);
      const s = spec();
      if (keyDocs[word] && /^[A-Za-z_]+:/.test(doc.lineAt(pos.line).text)) {
        return new vscode.Hover(new vscode.MarkdownString(`**${word}** — ${keyDocs[word]}`), range);
      }
      if (s.apis.includes(word)) {
        const note = s.apiNotes[word];
        const members = (s.apiMembers[word] ?? []).map((m) => `\`${word}.${m}\``).join(", ");
        return new vscode.Hover(new vscode.MarkdownString(`**${word}** — ${note ?? (members || "capability api")}`), range);
      }
      return undefined;
    },
  };
}

// ---- registration ---------------------------------------------------------

export function registerFrontmatter(context: vscode.ExtensionContext, getSpec: () => EvtSpec): void {
  const diagnostics = vscode.languages.createDiagnosticCollection("evt");
  context.subscriptions.push(
    diagnostics,
    vscode.languages.registerCompletionItemProvider(EVT_SELECTOR, completionProvider(getSpec), "-", " ", ":"),
    vscode.languages.registerHoverProvider(EVT_SELECTOR, hoverProvider(getSpec)),
    vscode.workspace.onDidOpenTextDocument((d) => validate(d, getSpec(), diagnostics)),
    vscode.workspace.onDidChangeTextDocument((e) => validate(e.document, getSpec(), diagnostics)),
    vscode.workspace.onDidCloseTextDocument((d) => diagnostics.delete(d.uri)),
  );
  for (const d of vscode.workspace.textDocuments) validate(d, getSpec(), diagnostics);
}

File: editors/nvim/lua/evaluate/frontmatter.lua
-- Frontmatter IntelliSense for `.evt` buffers: completion (omnifunc), diagnostics, and
-- hover — a Lua port of editors/vscode/src/frontmatter.ts. Reads the spec loaded by
-- evaluate/spec.lua (generated `evaluate-api.json`, with a bundled fallback), so the
-- keys/apis/hooks it offers and validates track the exact Godot + Evaluate in the project.

local spec_mod = require("evaluate.spec")

local M = {}

local NS = vim.api.nvim_create_namespace("evaluate.frontmatter")
local configs = {} -- bufnr -> resolved plugin config (for hover/lua_ls handoff)

-- ---- parsing (1-indexed lines, matching nvim_buf_get_lines rows +1) --------

-- The leading `--- … ---` block as { s = <opening row>, e = <closing row> } (1-indexed),
-- or nil when the file has no signature / an unterminated block.
local function fm_range(lines)
  if #lines == 0 or vim.trim(lines[1]) ~= "---" then return nil end
  for i = 2, #lines do
    if vim.trim(lines[i]) == "---" then return { s = 1, e = i } end
  end
  return nil
end

-- The nearest top-level `key:` at or above row `row` (the section being edited under).
local function section_at(lines, fm, row)
  for i = row, fm.s + 1, -1 do
    local k = lines[i]:match("^([%a_]+):")
    if k then return k end
  end
  return nil
end

-- Values already listed under `section` (so completion can skip them).
local function listed_under(lines, fm, section)
  local out, in_section = {}, false
  for i = fm.s + 1, fm.e - 1 do
    local key = lines[i]:match("^([%a_]+):")
    if key then
      in_section = key == section
    elseif in_section then
      local item = lines[i]:match("^%s*%-%s*([^:#%s]+)")
      if item then out[item] = true end
    end
  end
  return out
end

-- Node-attached scripts (`self`-bound, node hooks): `*.behavior.evt` behaviors (`.node.evt`
-- is its deprecated alias) and `*.statemachine.evt` state machines.
local function is_node_script(bufnr)
  local name = vim.api.nvim_buf_get_name(bufnr)
  return name:match("%.node%.evt$") ~= nil
    or name:match("%.behavior%.evt$") ~= nil
    or name:match("%.statemachine%.evt$") ~= nil
end

local function hooks_for(bufnr, spec)
  return is_node_script(bufnr) and spec.nodeHooks or spec.systemHooks
end

-- The capability apis this file declared in `apis:` (what the sandbox actually exposes).
local function declared_apis(lines, fm)
  return fm and listed_under(lines, fm, "apis") or {}
end

-- ---- completion (omnifunc) -------------------------------------------------

-- omnifunc protocol: (1, "") -> start column; (0, base) -> candidate list.
function M.omnifunc(findstart, base)
  local bufnr = vim.api.nvim_get_current_buf()
  local pos = vim.api.nvim_win_get_cursor(0)
  local row, col = pos[1], pos[2] -- row 1-indexed, col 0-indexed (byte before cursor)
  local line = vim.api.nvim_get_current_line()

  if findstart == 1 then
    local start = col
    while start > 0 and line:sub(start, start):match("[%w_]") do
      start = start - 1
    end
    M._start = start
    return start
  end

  local lines = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
  local fm = fm_range(lines)
  if not fm or not (row > fm.s and row < fm.e) then return {} end

  local spec = spec_mod.load(bufnr, configs[bufnr])
  local prefix = line:sub(1, col)
  local items = {}

  -- A list item under a section -> that section's values.
  if prefix:match("^%s*%-%s*%S*$") then
    local section = section_at(lines, fm, row)
    if section == "apis" then
      local used = listed_under(lines, fm, "apis")
      for _, a in ipairs(spec.apis) do
        if not used[a] then
          local members = spec.apiMembers[a] or {}
          local info = spec.apiNotes[a]
            or ("Members: " .. (#members > 0 and (a .. "." .. table.concat(members, ", " .. a .. ".")) or "(node)"))
          table.insert(items, { word = a, kind = "api", menu = "capability", info = info })
        end
      end
      -- 0.10.0 apis-as-modules: any Godot class/enum is declarable too, each injected as a
      -- bare global (only known when the live spec is loaded).
      local blocked = {}
      for _, b in ipairs(spec.blockedApis or {}) do blocked[b] = true end
      for _, cls in ipairs(spec.godotClasses or {}) do
        if not used[cls] and not blocked[cls] then
          table.insert(items, { word = cls, kind = "class", menu = "godot", info = "Godot class (injected as a bare global)" })
        end
      end
      for _, en in ipairs(spec.godotEnums or {}) do
        if not used[en] then
          table.insert(items, { word = en, kind = "enum", menu = "godot", info = "Godot global enum (injected as a bare global)" })
        end
      end
    elseif section == "register" then
      local used = listed_under(lines, fm, "register")
      local kind = is_node_script(bufnr) and "node hook" or "system hook"
      for _, h in ipairs(hooks_for(bufnr, spec)) do
        if not used[h] then table.insert(items, { word = h, kind = "hook", menu = kind }) end
      end
    end
    -- config/scenes/assets/returns/require: free text (no suggestions)
    return M._filter(items, base)
  end

  -- Start of a top-level line -> frontmatter keys.
  if prefix:match("^%s*[%a_]*$") and not prefix:match("^%s+") then
    local present = {}
    for i = fm.s + 1, fm.e - 1 do
      local k = lines[i]:match("^([%a_]+):")
      if k then present[k] = true end
    end
    for _, k in ipairs(spec.frontmatterKeys) do
      if not present[k] then
        table.insert(items, { word = k, kind = "key", menu = "frontmatter", info = spec_mod.KEY_DOCS[k] })
      end
    end
  end

  return M._filter(items, base)
end

function M._filter(items, base)
  if not base or base == "" then return items end
  local out = {}
  for _, it in ipairs(items) do
    if it.word:sub(1, #base) == base then table.insert(out, it) end
  end
  return out
end

-- ---- diagnostics -----------------------------------------------------------

local function escape(s)
  return (s:gsub("[%(%)%.%%%+%-%*%?%[%]%^%$]", "%%%1"))
end

-- Standalone-identifier byte positions (1-based start) of `w` in `line` — not preceded by a
-- word char / `.` / `:` and not followed by a word char.
local function ident_positions(line, w)
  local out, init, ew = {}, 1, escape(w)
  while true do
    local s, e = line:find(ew, init, true)
    if not s then break end
    local before = s > 1 and line:sub(s - 1, s - 1) or ""
    local after = line:sub(e + 1, e + 1)
    if (before == "" or not before:match("[%w_.:]")) and not after:match("[%w_]") then
      table.insert(out, s)
    end
    init = e + 1
  end
  return out
end

-- The next non-space char at/after byte index `i`.
local function next_token(line, i)
  local rest = line:sub(i):match("^%s*(.)")
  return rest or ""
end

-- Body references to a capability api the frontmatter did not declare. The sandbox only
-- exposes declared apis, so such a use is nil at runtime — flag it. A name the body locally
-- rebinds (`local save` / `save = …`) shadows the absent global, so it is excluded.
local function undeclared_api_uses(lines, body_start, all_apis, declared)
  local code = {}
  for i = 1, #lines do
    code[i] = i > body_start and lines[i]:gsub("%-%-.*$", "") or ""
  end
  local body = table.concat(code, "\n")

  local undeclared = {}
  for _, a in ipairs(all_apis) do
    if not declared[a] then
      local rebound = body:find("%f[%w_]local%s+" .. escape(a) .. "%f[^%w_]") ~= nil
      if not rebound then
        for _, s in ipairs(ident_positions(body, a)) do
          local nt = next_token(body, s + #a)
          if nt == "=" and body:sub(s + #a):match("^%s*=%s*[^=]") then
            rebound = true
            break
          end
        end
      end
      if not rebound then undeclared[a] = true end
    end
  end

  local diags = {}
  for i = body_start + 1, #lines do
    for a in pairs(undeclared) do
      for _, s in ipairs(ident_positions(code[i], a)) do
        local nt = next_token(code[i], s + #a)
        if nt == "." or nt == ":" or nt == "(" then
          table.insert(diags, {
            lnum = i - 1,
            col = s - 1,
            end_col = s - 1 + #a,
            severity = vim.diagnostic.severity.WARN,
            source = "evt",
            message = ("'%s' is used but not declared in 'apis:'. The sandbox only exposes declared "
              .. "apis, so this is nil at runtime. Add '%s' to 'apis:'."):format(a, a),
          })
        end
      end
    end
  end
  return diags
end

function M.validate(bufnr)
  if vim.bo[bufnr].filetype ~= "evt" then return end
  local lines = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
  local fm = fm_range(lines)
  if not fm then
    vim.diagnostic.reset(NS, bufnr)
    return
  end

  local spec = spec_mod.load(bufnr, configs[bufnr])
  local valid_hooks = {}
  for _, h in ipairs(hooks_for(bufnr, spec)) do valid_hooks[h] = true end
  local valid_apis = {}
  for _, a in ipairs(spec.apis) do valid_apis[a] = true end
  local valid_keys = {}
  for _, k in ipairs(spec.frontmatterKeys) do valid_keys[k] = true end
  local blocked_apis = {}
  for _, b in ipairs(spec.blockedApis or {}) do blocked_apis[b] = true end

  local diags, section = {}, nil
  for i = fm.s + 1, fm.e - 1 do
    local text = lines[i]
    local key = text:match("^([%a_]+):")
    if key then
      section = key
      if not valid_keys[section] then
        table.insert(diags, {
          lnum = i - 1, col = 0, end_col = #key, severity = vim.diagnostic.severity.WARN, source = "evt",
          message = ("Unknown frontmatter key '%s'. Valid: %s."):format(section, table.concat(spec.frontmatterKeys, ", ")),
        })
      end
    else
      local lead, value = text:match("^(%s*%-%s*)([^:#%s]+)")
      if value then
        local col = #lead
        if section == "apis" then
          -- The full token (incl. any ':'), to catch the removed `godot:Class` form.
          local entry = text:match("^%s*%-%s*(%S+)") or value
          if entry:match("^godot:") then
            local bare = entry:sub(#"godot:" + 1)
            table.insert(diags, {
              lnum = i - 1, col = col, end_col = col + #entry, severity = vim.diagnostic.severity.ERROR, source = "evt",
              message = ("'godot:'-prefixed entries were removed in 0.10.0 — declare the Godot class/enum "
                .. "name directly (e.g. '%s'); it is injected as a bare global."):format(bare ~= "" and bare or "Node3D"),
            })
          elseif blocked_apis[value] or value:match("^InputEvent") then
            local is_input = value:match("^Input") or value:match("^Key")
              or value:match("^Joy") or value:match("^Mouse")
            local why = is_input
              and "raw input is native-only — subscribe to mapped actions via the 'actions' api"
              or "assets load via frontmatter assets:; persistence via save/sql"
            table.insert(diags, {
              lnum = i - 1, col = col, end_col = col + #value, severity = vim.diagnostic.severity.ERROR, source = "evt",
              message = ("'%s' is blocked from declaration — %s."):format(value, why),
            })
          elseif not valid_apis[value] and not value:match("^%u") then
            -- PascalCase entries are Godot classes/enums (or host extension apis) — accepted.
            table.insert(diags, {
              lnum = i - 1, col = col, end_col = col + #value, severity = vim.diagnostic.severity.WARN, source = "evt",
              message = ("Unknown api '%s'. Framework services: %s; Godot classes/enums are declared by "
                .. "their PascalCase name. (It must be declared to be usable.)"):format(
                value, table.concat(spec.apis, ", ")),
            })
          end
        elseif section == "register" and not valid_hooks[value] then
          table.insert(diags, {
            lnum = i - 1, col = col, end_col = col + #value, severity = vim.diagnostic.severity.WARN, source = "evt",
            message = ("Unknown %s hook '%s'. Valid: %s."):format(
              is_node_script(bufnr) and "node" or "system", value, table.concat(hooks_for(bufnr, spec), ", ")),
          })
        elseif section == "assets" and not text:match("^%s*%-%s*[%a_][%w_]*%s*:") then
          -- 0.10.0: assets are a name -> path map (or list of single-key maps); bare paths error.
          table.insert(diags, {
            lnum = i - 1, col = col, end_col = col + #value, severity = vim.diagnostic.severity.ERROR, source = "evt",
            message = ("Bare-path 'assets:' entries are an error since 0.10.0 — name it ('<name>: \"%s\"'), "
              .. "injected as 'assets.<name>'."):format(value),
          })
        end
      end
    end
  end

  vim.list_extend(diags, undeclared_api_uses(lines, fm.e, spec.apis, declared_apis(lines, fm)))
  vim.diagnostic.set(NS, bufnr, diags)
end

-- ---- hover -----------------------------------------------------------------

-- Markdown lines describing the frontmatter word under the cursor, or nil (so a `K` mapping
-- can fall back to the Lua-body LSP for words in the body).
function M.hover_lines(bufnr)
  local lines = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
  local fm = fm_range(lines)
  local pos = vim.api.nvim_win_get_cursor(0)
  local row = pos[1]
  if not fm or not (row > fm.s and row < fm.e) then return nil end

  local word = vim.fn.expand("<cword>")
  if word == "" then return nil end
  local spec = spec_mod.load(bufnr, configs[bufnr])

  local line = lines[row]
  if spec_mod.KEY_DOCS[word] and line:match("^[%a_]+:") then
    return { ("**%s** — %s"):format(word, spec_mod.KEY_DOCS[word]) }
  end
  if vim.tbl_contains(spec.apis, word) then
    local note = spec.apiNotes[word]
    local members = spec.apiMembers[word] or {}
    local detail = note
      or (#members > 0 and ("`" .. word .. "." .. table.concat(members, "`, `" .. word .. ".") .. "`") or "capability api")
    return { ("**%s** — %s"):format(word, detail) }
  end
  if vim.tbl_contains(spec.blockedApis or {}, word) then
    return { ("**%s** — blocked from `apis:` — assets load via frontmatter `assets:`."):format(word) }
  end
  if vim.tbl_contains(spec.godotClasses or {}, word) then
    return { ("**%s** — Godot class. Declared in `apis:`, injected as a bare global (`%s.new()`, statics, enums)."):format(word, word) }
  end
  if vim.tbl_contains(spec.godotEnums or {}, word) then
    return { ("**%s** — Godot global enum. Declared in `apis:`, injected as a bare global (`%s.<Constant>`)."):format(word, word) }
  end
  return nil
end

-- ---- attach ----------------------------------------------------------------

function M.attach(bufnr, config)
  configs[bufnr] = config
  vim.bo[bufnr].omnifunc = "v:lua.require'evaluate.frontmatter'.omnifunc"

  local group = vim.api.nvim_create_augroup("evaluate.frontmatter." .. bufnr, { clear = true })
  vim.api.nvim_create_autocmd({ "TextChanged", "TextChangedI", "InsertLeave", "BufWritePost", "BufEnter" }, {
    group = group,
    buffer = bufnr,
    callback = function() M.validate(bufnr) end,
  })
  vim.api.nvim_create_autocmd("BufUnload", {
    group = group,
    buffer = bufnr,
    callback = function()
      configs[bufnr] = nil
      vim.diagnostic.reset(NS, bufnr)
    end,
  })

  if config and config.hover_key and config.hover_key ~= "" then
    vim.keymap.set("n", config.hover_key, function()
      local hl = M.hover_lines(bufnr)
      if hl then
        vim.lsp.util.open_floating_preview(hl, "markdown", { border = "rounded", focus = false })
      else
        vim.lsp.buf.hover() -- body word: defer to the Lua LSP
      end
    end, { buffer = bufnr, desc = "Evaluate: hover (frontmatter or Lua body)" })
  end

  M.validate(bufnr)
end

return M
